### PR TITLE
add AXFR source to OctoDNS

### PR DESCRIPTION
**Pull request by yzguy**
Adds a new source requested in #239. This source allows a user to pull data from
a legacy system (Bind9, etc.) that does not have an API/existing provider via
AXFR Zone Transfer.

I tested it against NS1 via AXFR with a zone with ~6500 records and it worked like a dream, might be missing some record types, but it's a good start. Figured I'd get it up here for some review while I tweak/add tests.

Ideally, it could be highly configurable:
* TSIG
* Different Port
* Timeout

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -162,7 +162,8 @@ The above command pulled the existing data out of Route53 and placed the results
 | [PowerDnsProvider](/octodns/provider/powerdns.py) | | All | No | |
 | [Rackspace](/octodns/provider/rackspace.py) | | A, AAAA, ALIAS, CNAME, MX, NS, PTR, SPF, TXT | No |  |
 | [Route53](/octodns/provider/route53.py) | boto3 | A, AAAA, CAA, CNAME, MX, NAPTR, NS, PTR, SPF, SRV, TXT | Yes | |
-| [TinyDNSSource](/octodns/source/tinydns.py) | | A, CNAME, MX, NS, PTR | No | read-only |
+| [TinyDnsFileSource](/octodns/source/tinydns.py) | | A, CNAME, MX, NS, PTR | No | read-only |
+| [AxfrSource](/octodns/source/axfr.py) | | A, AAAA, CNAME, MX, NS, PTR, SPF, SRV, TXT | No | read-only |
```
Comment (ross) [inline README.md]: The rest of these are in alphabetical order. Probably good to move this up to the top. We could break things out into two tables, providers and sources, if you'd prefer to group them.

File: octodns/source/axfr.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                                                   relativize=False),
+                                     relativize=False)
+        except DNSException:
+            raise AxfrSourceZoneTransferFailed()
+
+        records = []
+
+        for (name, ttl, rdata) in zone.iterate_rdatas():
+            rdtype = dns.rdatatype.to_text(rdata.rdtype)
+            if rdtype != 'SOA':
```
Comment (ross) [inline octodns/source/axfr.py]: Should this use `SUPPORTS` to filter things to only the supported record types?

Comment (yzguy) [reply octodns/source/axfr.py]: I was thinking about this today, I think I would end up removing it here, and adding it within the populate method.

Ideally this `zone_records` method is just the thing that performs the zone transfer and transforms the response into a more sane data structure, it shouldn't care about filtering records by type.

However `populate` should care about what record types get let through.

I will make the changes.

**Comment by ross** [comment]
Just checking that you're ready for this to :ship: before I merge...

**Comment by yzguy** [comment]
Need to write the tests for it, it'll be a little confusing because I have to probably mock the response from the zone transfer, which is a bunch of dnspython objects, etc.

I'm not sure if this is the best spot to put it, but I added a ZoneFileSource as well, as the only difference with loading from a zone file or a zone transfer with dnspython is a method, so it was a quick win to add a source for a zone file.

It's a little confusing it being under the axfr module, but I'm not sure what would be the best place to put it. I'm open to any suggestions on better placement/organization.

**Comment by yzguy** [comment]
Added test, instead of creating a bunch of dnspython objects, I created a small zone file, then load it via dnspython's `from_file` method, and then just return that when `from_xfr` is called. Both methods generate a dns.zone.Zone object.

**Pull request merged**